core/queue/queue.go

```diff
@@ -0,0 +1,313 @@
+			return
+		}
+		if errors.Is(err, errRemoved) {
+			continue
+		}
+		if err != nil {
+			q.logger.Error("handle next", zap.Error(err))
+		}
+		if q.options.handlerTickPeriod > 0 {
+			select {
+			case <-time.After(q.options.handlerTickPeriod):
+			case <-q.ctx.Done():
+				return
+			}
+		}
+	}
+
+}
+
+func (q *Queue[T]) handleNext() error {
+	it, err := q.batcher.WaitOne(q.ctx)
+	if err != nil {
+		return fmt.Errorf("wait one: %w", err)
+	}
+	ok := q.Has(it.Key())
+	if !ok {
+		return errRemoved
+	}
+
+	action, err := q.handler(q.ctx, it)
+	atomic.AddUint32(&q.handledItems, 1)
+	switch action {
+	case ActionDone:
+		removeErr := q.Remove(it.Key())
+		if removeErr != nil {
+			return fmt.Errorf("remove from queue: %w", removeErr)
+		}
+	case ActionRetry:
+		addErr := q.batcher.Add(q.ctx, it)
+		if addErr != nil {
+			return fmt.Errorf("add to queue: %w", addErr)
+		}
+	}
+	if err != nil {
+		return fmt.Errorf("handler: %w", err)
+	}
+	return nil
+}
+
+func (q *Queue[T]) restore() error {
+	items, err := q.listItemsByPrefix()
+	if err != nil {
+		return fmt.Errorf("list items: %w", err)
+	}
+	err = q.batcher.Add(q.ctx, items...)
+	if err != nil {
+		return fmt.Errorf("add to queue: %w", err)
+	}
+	for _, it := range items {
+		q.set[it.Key()] = struct{}{}
+	}
+	return nil
+}
+
+func (q *Queue[T]) listItemsByPrefix() ([]T, error) {
+	var items []T
+	err := q.db.View(func(txn *badger.Txn) error {
+		it := txn.NewIterator(badger.IteratorOptions{
+			PrefetchSize:   100,
+			PrefetchValues: true,
+			Prefix:         q.badgerPrefix,
+		})
+		defer it.Close()
+
+		for it.Rewind(); it.Valid(); it.Next() {
+			item := it.Item()
+			qItem, err := q.unmarshalItem(item)
+			if err != nil {
+				return fmt.Errorf("get queue item %s: %w", item.Key(), err)
+			}
+			items = append(items, qItem)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(items) > 0 {
+		var itemIface Item = items[0]
+		if _, ok := itemIface.(ItemWithOrder); ok {
+			sort.Slice(items, func(i, j int) bool {
+				var left Item = items[i]
+				var right Item = items[j]
+				return left.(ItemWithOrder).Less(right.(ItemWithOrder))
+			})
+		}
+	}
+
+	return items, nil
+}
+
+func (q *Queue[T]) unmarshalItem(item *badger.Item) (T, error) {
+	it := q.factoryFunc()
+	err := item.Value(func(raw []byte) error {
+		return json.Unmarshal(raw, it)
+	})
+	if err != nil {
+		return it, err
+	}
+	return it, nil
+}
+
+func (q *Queue[T]) Close() error {
+	q.ctxCancel()
+	return q.batcher.Close()
+}
+
+func (q *Queue[T]) makeKey(itemKey string) []byte {
+	return append(q.badgerPrefix, []byte(itemKey)...)
+}
+
+func (q *Queue[T]) Add(item T) error {
+	err := q.checkClosed()
+	if err != nil {
+		return err
+	}
+
+	q.lock.Lock()
+	if _, ok := q.set[item.Key()]; ok {
+		q.lock.Unlock()
+		return nil
+	}
+	q.set[item.Key()] = struct{}{}
+	q.lock.Unlock()
+
+	err = q.batcher.Add(q.ctx, item)
+	if err != nil {
+		return err
+	}
+	return q.store(item)
```

Comment:
I would split all the badger parts into other components, maybe just provide an interface inside queue which would store the items, and remove them. And make this interface abstract